### PR TITLE
Add action server

### PR DESCRIPTION
**Pull request by Abishalini**
Adding an action server so that we can stall the calibration process until we get a goal from client (in our case, from MoveIt Studio agent).

The goal will contain the initial estimated pose for the camera (and maybe the checkerboard too)
If the estimated pose is not sent from the client, the initial poses is read from `calibrate.yaml` 
The goal can also specify if we should lookup the initial position of the camera from the TF buffer (Enable the lookup_estimate flag). We currently cannot update the checkerboard frame this way since we don't specify this frame in the URDF.

Once the calibration routine is done, the offsets are then sent as a result. 

TODO:
- [ ] The node needs to brought down and up again between each client call since some parameters get declared again which causes the program to crash. Need to fix
- [ ] Do not comment out the spin_some node calls. This will break the routine when we don't use the action server.
- [ ] Right now, the offsets/results from the calibration process are not correct. Need to fix

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (MarqRazz) [inline .vscode/settings.json]: Whats this file for? I keep my code settings in the base of the workspace so we don' track files like this.

File: robot_calibration_msgs/action/CalibratePose.action [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+geometry_msgs/TransformStamped[] estimated_poses
+
+bool lookup_initial_estimate
+
+# where should we save the results? If not specified then what?
+# should this be a separate service?
+string result_output_path
+
+# non-empty path activates this feature
+string bag_file_path
```
Comment (MarqRazz) [inline robot_calibration_msgs/action/CalibratePose.action]: I vote we remove these.

**Comment by MarqRazz** [comment]
I still need to write a description but here's my [action client](https://github.com/PickNikRobotics/moveit_studio/pull/1596).

File: .vscode/settings.json (hunk 1 above)
Comment (Abishalini) [reply .vscode/settings.json]: Removed

File: robot_calibration/include/robot_calibration/util/action_client.hpp [hunk 3]
```diff
@@ -91,7 +91,7 @@ class ActionClient
     rclcpp::Time start = node->now();
     while (state_ == ActionClientState::ACTIVE)
     {
-      rclcpp::spin_some(node);
+      // rclcpp::spin_some(node);
```
Comment (MarqRazz) [inline robot_calibration/include/robot_calibration/util/action_client.hpp:94]: Won't removing these break the other interface?

File: robot_calibration/src/nodes/calibrate_server.cpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  // Get calibration poses YAML path
+  std::string data_source("calibration_poses.yaml");
+  if (argc > 1)
+  {
+    data_source = argv[1];
+  }
+
+  std::cout << "Creating action server" << std::endl;
+  CalibratePoseServer calibrate_pose_action_server = CalibratePoseServer(node, data_source);
+
```
Comment (MarqRazz) [inline robot_calibration/src/nodes/calibrate_server.cpp:21]: Could you add this comment here?
`// TODO: This's a workaround a bug in MultiThreadedExecutor deadlock when spinning without a timeout`

File: robot_calibration/src/util/action_server.cpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+    std::cout << "Parameter Offsets:" << std::endl;
+    std::cout << opt.getOffsets()->getOffsetYAML() << std::endl;
+  }
+
+  // Write outputs
+  // robot_calibration::exportResults(opt, description_msg.data, data);
+  auto result = std::make_shared<CalibratePose::Result>();
+
+  RCLCPP_INFO(logger, "Done calibrating");
```
Comment (MarqRazz) [inline robot_calibration/src/util/action_server.cpp:208]: Can we add the result with a TODO that the pose's need to be filled in
```
// If we make it here, running the calibration succeeded
auto calibrate_action_result = std::make_shared<CalibratePose::Result>();
// calibrate_action_result->calibrated_poses.push_back(frame);  // TODO fill in the answer
has_active_action_goal_ = false;
goal_handle->succeed(calibrate_action_result);
```

**Review by MarqRazz — APPROVED** [review]

File: robot_calibration/src/util/action_server.cpp (hunk 5 above)
Comment (Abishalini) [reply robot_calibration/src/util/action_server.cpp:208]: Added the result from the cal process now

File: robot_calibration/src/nodes/calibrate_server.cpp (hunk 4 above)
Comment (Abishalini) [reply robot_calibration/src/nodes/calibrate_server.cpp:21]: What is the difference between `rclcpp::spin(node)` and using a MultiThreadedExecutor? Both seems to work fine

File: robot_calibration/include/robot_calibration/util/action_client.hpp (hunk 3 above)
Comment (Abishalini) [reply robot_calibration/include/robot_calibration/util/action_client.hpp:94]: It does break the other interface but keeping the `spin_some` calls will throw this error. 
```
[calibrate_server-1] terminate called after throwing an instance of 'std::runtime_error'
[calibrate_server-1]   what():  Node '/robot_calibration' has already been added to an executor.

```
We will have to remove spinning the node in the main to enable all the `spin_some` calls. Let me see what can be done

File: robot_calibration/src/nodes/calibrate_server.cpp (hunk 4 above)
Comment (MarqRazz) [reply robot_calibration/src/nodes/calibrate_server.cpp:21]: In the code below we add the node (CalibratePoseServer) to the multi-threaded executor which is capable of spinning a multi-threaded node (when built correctly the CalibratePoseServer can create as many thread as it requires to listen for topics/services/actions).

When you call `spin(node)` you are calling in on the node so it depends what type of node you have built. If it node has listeners configured with:
`node->create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive)` or
`node->create_callback_group(rclcpp::CallbackGroupType::Reentrant)`
then the nodes executor needs to support multi-threading.

**Pull request merged**